### PR TITLE
CharUnicodeInfo optimizations + flow TR-29 text segmentation through

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
This PR involves three general poles.

 1. Rewrite `GenUnicodeProp` to include more Unicode data, including UAX29 text segmentation info and optional case folding information. It also includes data that was previously missing.
 2. Rewrite `CharUnicodeInfo` to be a bit more optimized, including cramming more data into the existing tables + arrays.
 3. Update `StringInfo` and `TextElementEnumerator` to understand Unicode UAX29 grapheme cluster boundaries for segmentation purposes.

When reviewing, I'd probably go in the order below.

### `tools/GenUnicodeProp`

This is the logic that reads all of the Unicode data files and spits out the `CharUnicodeInfoData.cs` backing file. Much of this application was rewritten to remove unneeded logic and to add support for smuggling complex data through the 11:5:4 tables.

There were also some bug fixes made to this. For instance, the previous incarnation of this tool looked only at [`UnicodeData.txt`](https://www.unicode.org/Public/12.1.0/ucd/UnicodeData.txt) when extracting data. However, some necessary data also exists in ancillary files such as [`DerivedBidiClass.txt`](https://www.unicode.org/Public/12.1.0/ucd/extracted/DerivedBidiClass.txt). One example of this is the code point `U+1EEEF`, which does not appear in `UnicodeData.txt` because it is not yet assigned, but whose bidi category should be "AL" (strongly right-to-left) because of the block in which it resides. Previous versions of `GenUnicodeProp` would emit incorrect data for code points such as these.

### `CharUnicodeInfo.cs`

The methods in this file were rewritten to be more optimized, and methods were moved around a bit so that logical method groups (such as `GetUnicodeCategory`) stay together. Of note is that there are many places where a single table or a single array holds multiple pieces of information.

For example, the _Categories_ 11:5:4 table is used to index into the `CategoriesValues` array, where each `byte` element of that array is essentially a glorified `(isWhiteSpace, bidiCategory, unicodeCategory)` tuple. The _NumericGrapheme_ 11:5:4 table is used to index into the `DigitValues`, `NumericValues`, and `GraphemeSegmentationValues` arrays.

### `StrongBidiCategory.cs` and `IdnMapping.cs`

The previous incarnation of `CharUnicodeInfo` held full bidi class data. However, it turns out that we don't have a public API to get at this information, and the only consumer of this information in the entire framework is `IdnMapping`. Furthermore, it doesn't care about most bidi classes. It only cares about classes that are strongly left-to-right ("L") or strongly right-to-left ("R", "AL"). So I changed the data stored in `CharUnicodeInfo` to reflect only this very limited set of information rather than the full bidi class. See https://www.unicode.org/reports/tr44/#BC_Values_Table for more info on these values.

When generating the backing table (see the `GenUnicodeProp` tool), all code points marked with "L" are given a "strong left-to-right" marker in the backing data, and all code points marked with "AL" or "R" are given a "strong right-to-left" marker in the backing data. All other bidi information is thrown away when generating `CharUnicodeInfoData.cs`.

### `Rune.cs` and `Char.cs`

Slight modifications to these files to react to refactorings in `CharUnicodeInfo`.

### `TextSegmentationUtility.cs`

This is the workhorse class that computes grapheme cluster segmentation boundaries ("display characters", essentially). We use the definition of "extended grapheme cluster" per https://www.unicode.org/reports/tr29/. See Sec. 3.1 and 3.1.1 of that document for the specific algorithm we follow.

The code is generic so that it can work with either UTF-16 or UTF-8. Currently only the UTF-16 API is public, implemented by passing the delegate [`Rune.DecodeFromUtf16`](https://apisof.net/catalog/System.Text.Rune.DecodeFromUtf16(ReadOnlySpan%3CChar%3E,Rune,Int32)) down to the workhorse routine. If we wanted to add UTF-8 support in the future, it would be trivial to do so by instead calling the same workhorse with the [`Rune.DecodeFromUtf8`](https://apisof.net/catalog/System.Text.Rune.DecodeFromUtf8(ReadOnlySpan%3CByte%3E,Rune,Int32)) delegate.

### `StringInfo.cs` and `TextElementEnumerator.cs`

Rewritten to use the new "extended grapheme cluster" logic in `TextSegmentationUtility`.

There are __no public API changes__ as part of this work. I haven't yet written / updated unit tests, so everything will fail until those tests are available. Fortunately, there's some Unicode test data available at https://www.unicode.org/Public/12.1.0/ucd/auxiliary/GraphemeBreakTest.txt, so I'll just pull that in directly and use it as part of the test suite.

### CODE REVIEW

**Comment by GrabYourPitchforks** [comment]
Some very preliminary perf results follow. (_Value_ is the decimal representation of the code point we're passing in.)

|               Method | Toolchain |  Value |     Mean |     Error |    StdDev |   Median | Ratio | RatioSD |
|--------------------- |---------- |------- |---------:|----------:|----------:|---------:|------:|--------:|
| **GetDecimalDigitValue** |    **master** |      **0** | **2.473 ns** | **0.0417 ns** | **0.0390 ns** | **2.465 ns** |  **1.00** |    **0.00** |
| GetDecimalDigitValue |     proto |      0 | 2.479 ns | 0.0817 ns | 0.0764 ns | 2.453 ns |  1.00 |    0.04 |
|                      |           |        |          |           |           |          |       |         |
|   GetUnicodeCategory |    master |      0 | 3.076 ns | 0.0941 ns | 0.1006 ns | 3.014 ns |  1.00 |    0.00 |
|   GetUnicodeCategory |     proto |      0 | 2.506 ns | 0.0689 ns | 0.0645 ns | 2.477 ns |  0.81 |    0.04 |
|                      |           |        |          |           |           |          |       |         |
| **GetDecimalDigitValue** |    **master** |  **55296** | **2.492 ns** | **0.0831 ns** | **0.0816 ns** | **2.500 ns** |  **1.00** |    **0.00** |
| GetDecimalDigitValue |     proto |  55296 | 2.614 ns | 0.0913 ns | 0.0810 ns | 2.625 ns |  1.05 |    0.05 |
|                      |           |        |          |           |           |          |       |         |
|   GetUnicodeCategory |    master |  55296 | 2.715 ns | 0.0870 ns | 0.1036 ns | 2.691 ns |  1.00 |    0.00 |
|   GetUnicodeCategory |     proto |  55296 | 2.578 ns | 0.0835 ns | 0.1225 ns | 2.554 ns |  0.94 |    0.06 |
|                      |           |        |          |           |           |          |       |         |
| **GetDecimalDigitValue** |    **master** | **128512** | **2.448 ns** | **0.0781 ns** | **0.0802 ns** | **2.460 ns** |  **1.00** |    **0.00** |
| GetDecimalDigitValue |     proto | 128512 | 2.645 ns | 0.0778 ns | 0.0727 ns | 2.656 ns |  1.08 |    0.04 |
|                      |           |        |          |           |           |          |       |         |
|   GetUnicodeCategory |    master | 128512 | 3.398 ns | 0.2303 ns | 0.6792 ns | 3.109 ns |  1.00 |    0.00 |
|   GetUnicodeCategory |     proto | 128512 | 2.326 ns | 0.0389 ns | 0.0345 ns | 2.323 ns |  0.80 |    0.07 |

Looks like some methods aren't being properly inlined, which is causing the prototype not to be as optimal as it otherwise could be. I'll bring this up with the JIT folks.

File: src/System.Private.CoreLib/shared/System/Globalization/CharUnicodeInfo.cs [hunk 1]
```diff
@@ -71,251 +74,372 @@ internal static int InternalConvertToUtf32(StringBuilder s, int index)
                     if ((uint)temp2 <= HIGH_SURROGATE_RANGE)
                     {
                         // Convert the surrogate to UTF32 and get the result.
-                        return (temp1 * 0x400) + temp2 + UNICODE_PLANE01_START;
+                        c = (temp1 << 18) + temp2 + UNICODE_PLANE01_START;
```
Comment (tarekgh) [inline src/System.Private.CoreLib/shared/System/Globalization/CharUnicodeInfo.cs]: >c = (temp1 << 18) [](start = 24, length = 17)

could you please add a small comment telling this equivalent to multiplying by 0x400?

File: src/System.Private.CoreLib/shared/System/Globalization/CharUnicodeInfo.cs [hunk 2]
```diff
@@ -71,251 +74,372 @@ internal static int InternalConvertToUtf32(StringBuilder s, int index)
...
-                return BitConverter.Int64BitsToDouble(BinaryPrimitives.ReverseEndianness(Unsafe.ReadUnaligned<long>(ref value)));
-            }
-            return -1;
+        public static int GetDecimalDigitValue(char ch)
+        {
+            return GetDecimalDigitValueInternalNoBoundsCheck(ch);
         }
 
-        internal static byte InternalGetDigitValues(int ch, int offset)
+        public static int GetDecimalDigitValue(string s, int index)
```
Comment (tarekgh) [inline src/System.Private.CoreLib/shared/System/Globalization/CharUnicodeInfo.cs:110]: >string [](start = 47, length = 6)

should we make this nullable?

**Comment by tarekgh** [comment]
>         private static void AssertCategoryCasingTableLevels(int level1BitCount, int level2BitCount, int level3BitCount)

does this need to be here?

---
Refers to: src/System.Private.CoreLib/shared/System/Globalization/CharUnicodeInfoData.cs:15 in 0eb8313. [](commit_id = 0eb8313a6e0b91ac9a349bc1ca22b51abc00dc3d, deletion_comment = False)

File: src/System.Private.CoreLib/shared/System/Globalization/CharUnicodeInfo.cs (hunk 2 above)
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/shared/System/Globalization/CharUnicodeInfo.cs:110]: None of the other APIs allow null parameters - seems like allowing this to be nullable would be an outlier?

**Pull request closed without merging**